File: python/classes/recipe_manager.py
"""Recipe Creator and Manager — compositional entities, scaling, pantry checks, shopping lists.

A recipe is the first entity in the Rosetta Stone whose primary field is a
structured collection (an ordered list of line items), not atomic scalars.
Operations over recipes are therefore operations over that structure:

  * scale — a pure linear map over every line's quantity.
  * can_make — conjunction over N ingredients (width determined by recipe).
  * shopping_list — the delta between required and available.
  * make — atomic multi-entity consumption of pantry stock.

Units are matched strictly: if the pantry has 500 g of flour and the recipe
calls for 2 cups, the system reports incompatible units rather than guessing
a conversion. Unit algebra is an adjacent problem, not this problem.
"""

from dataclasses import dataclass, field
from typing import Optional
from fractions import Fraction


# Represent quantities as Fractions so scaling by 0.5 or 1/3 stays exact.
Qty = Fraction


def qty(value) -> Qty:
    """Construct a quantity from int/float/str/Fraction."""
    return value if isinstance(value, Fraction) else Fraction(str(value))


@dataclass
class Ingredient:
    id: str
    name: str
    default_unit: str = ""


@dataclass
class RecipeLine:
    """One ingredient requirement within a recipe."""
    ingredient_id: str
    quantity: Qty
    unit: str

    def scaled(self, factor: Qty) -> "RecipeLine":
        return RecipeLine(self.ingredient_id, self.quantity * factor, self.unit)


@dataclass
class Recipe:
    id: str
    name: str
    lines: list[RecipeLine] = field(default_factory=list)
    serves: int = 1
    steps: list[str] = field(default_factory=list)

    def scaled(self, factor: Qty) -> "Recipe":
        """A new recipe with every line's quantity multiplied by factor."""
        return Recipe(
            id=self.id, name=self.name,
            lines=[l.scaled(factor) for l in self.lines],
            serves=self.serves,  # serves is original; caller interprets
            steps=list(self.steps),
        )


@dataclass
class PantryItem:
    quantity: Qty
    unit: str


@dataclass
class Shortage:
    """One line's worth of missing stock, as a computed delta."""
    ingredient_id: str
    required: Qty
    available: Qty
    unit: str
    reason: str = "insufficient"  # or "unit_mismatch" or "not_stocked"

    @property
    def missing(self) -> Qty:
        return self.required - self.available


class RecipeError(Exception):
    pass


class UnknownEntity(RecipeError): ...
class UnitMismatch(RecipeError): ...
class InsufficientStock(RecipeError): ...


class RecipeBook:
    """Catalogue of ingredients and recipes, with a mutable pantry."""

    def __init__(self) -> None:
        self.ingredients: dict[str, Ingredient] = {}
        self.recipes: dict[str, Recipe] = {}
        self.pantry: dict[str, PantryItem] = {}

    # --- registration ---
    def add_ingredient(self, ingredient: Ingredient) -> None:
        if ingredient.id in self.ingredients:
            raise ValueError(f"ingredient already registered: {ingredient.id}")
        self.ingredients[ingredient.id] = ingredient

    def add_recipe(self, recipe: Recipe) -> None:
        if recipe.id in self.recipes:
            raise ValueError(f"recipe already registered: {recipe.id}")
        for line in recipe.lines:
            if line.ingredient_id not in self.ingredients:
                raise UnknownEntity(f"unknown ingredient: {line.ingredient_id}")
        self.recipes[recipe.id] = recipe

    def stock(self, ingredient_id: str, quantity, unit: str) -> None:
        """Set (not add) the pantry amount for an ingredient."""
        if ingredient_id not in self.ingredients:
            raise UnknownEntity(f"unknown ingredient: {ingredient_id}")
        self.pantry[ingredient_id] = PantryItem(qty(quantity), unit)

    def add_stock(self, ingredient_id: str, quantity, unit: str) -> None:
        """Increase pantry stock; fails on unit mismatch."""
        if ingredient_id not in self.ingredients:
            raise UnknownEntity(f"unknown ingredient: {ingredient_id}")
        q = qty(quantity)
        existing = self.pantry.get(ingredient_id)
        if existing is None:
            self.pantry[ingredient_id] = PantryItem(q, unit)
            return
        if existing.unit != unit:
            raise UnitMismatch(
                f"{ingredient_id}: pantry in {existing.unit}, adding {unit}")
        existing.quantity += q

    # --- scaling ---
    def scale(self, recipe_id: str, factor) -> Recipe:
        if recipe_id not in self.recipes:
            raise UnknownEntity(f"unknown recipe: {recipe_id}")
        return self.recipes[recipe_id].scaled(qty(factor))

    def _required_lines(self, recipe_id: str, servings: Optional[int]) -> list[RecipeLine]:
        if recipe_id not in self.recipes:
            raise UnknownEntity(f"unknown recipe: {recipe_id}")
        recipe = self.recipes[recipe_id]
        if servings is None or servings == recipe.serves:
            return list(recipe.lines)
        factor = Fraction(servings, recipe.serves)
        return [l.scaled(factor) for l in recipe.lines]

    # --- pantry evaluation ---
    def can_make(self, recipe_id: str, servings: Optional[int] = None
                 ) -> tuple[bool, list[Shortage]]:
        """Conjunction over recipe lines: all must be satisfiable from the pantry."""
        shortages: list[Shortage] = []
        for line in self._required_lines(recipe_id, servings):
            p = self.pantry.get(line.ingredient_id)
            if p is None:
                shortages.append(Shortage(
                    line.ingredient_id, line.quantity, Fraction(0),
                    line.unit, reason="not_stocked"))
                continue
            if p.unit != line.unit:
                shortages.append(Shortage(
                    line.ingredient_id, line.quantity, Fraction(0),
                    line.unit, reason="unit_mismatch"))
                continue
            if p.quantity < line.quantity:
                shortages.append(Shortage(
                    line.ingredient_id, line.quantity, p.quantity, line.unit))
        return (not shortages, shortages)

    def shopping_list(self, recipe_id: str, servings: Optional[int] = None
                      ) -> list[RecipeLine]:
        """Gap report: required minus available, per line. Only lines needing more appear."""
        out: list[RecipeLine] = []
        for line in self._required_lines(recipe_id, servings):
            p = self.pantry.get(line.ingredient_id)
            if p is None or p.unit != line.unit:
                out.append(RecipeLine(line.ingredient_id, line.quantity, line.unit))
                continue
            deficit = line.quantity - p.quantity
            if deficit > 0:
                out.append(RecipeLine(line.ingredient_id, deficit, line.unit))
        return out

    def make(self, recipe_id: str, servings: Optional[int] = None) -> Recipe:
        """Atomically consume all required ingredients. All-or-nothing."""
        ok, shortages = self.can_make(recipe_id, servings)
        if not ok:
            raise InsufficientStock(
                f"cannot make {recipe_id}: {len(shortages)} shortage(s): "
                f"{[(s.ingredient_id, s.reason) for s in shortages]}")
        for line in self._required_lines(recipe_id, servings):
            self.pantry[line.ingredient_id].quantity -= line.quantity
        return self.recipes[recipe_id]

    # --- queries over the recipe graph ---
    def recipes_using(self, ingredient_id: str) -> list[Recipe]:
        if ingredient_id not in self.ingredients:
            raise UnknownEntity(f"unknown ingredient: {ingredient_id}")
        return [r for r in self.recipes.values()
                if any(l.ingredient_id == ingredient_id for l in r.lines)]

    def ingredients_used(self) -> set[str]:
        """Every ingredient referenced by any recipe."""
        used: set[str] = set()
        for r in self.recipes.values():
            for l in r.lines:
                used.add(l.ingredient_id)
        return used


# --- demo ---
if __name__ == "__main__":
    book = RecipeBook()
    for i in [
        Ingredient("flour",  "all-purpose flour", "g"),
        Ingredient("sugar",  "white sugar",       "g"),
        Ingredient("butter", "unsalted butter",   "g"),
        Ingredient("egg",    "large egg",         "whole"),
        Ingredient("vanilla","vanilla extract",   "tsp"),
    ]:
        book.add_ingredient(i)

    cookies = Recipe(
        id="cookies", name="Sugar Cookies", serves=24,
        lines=[
            RecipeLine("flour",   qty(300), "g"),
            RecipeLine("sugar",   qty(200), "g"),
            RecipeLine("butter",  qty(225), "g"),
            RecipeLine("egg",     qty(2),   "whole"),
            RecipeLine("vanilla", qty(1),   "tsp"),
        ],
        steps=["cream butter and sugar", "add egg and vanilla",
               "fold in flour", "bake at 180C for 12 min"],
    )
    book.add_recipe(cookies)

    # Pantry has some, not all.
    book.stock("flour",  500, "g")
    book.stock("sugar",  150, "g")        # short on sugar
    book.stock("butter", 250, "g")
    book.stock("vanilla", 2,  "tsp")
    # egg is not stocked at all.

    ok, shortages = book.can_make("cookies")
    print(f"can make 24 cookies? {ok}")
    for s in shortages:
        print(f"  - {s.ingredient_id}: need {s.required}{s.unit}, "
              f"have {s.available}{s.unit} ({s.reason})")

    print("\nshopping list for 24 cookies:")
    for line in book.shopping_list("cookies"):
        print(f"  buy {line.quantity} {line.unit} of {line.ingredient_id}")

    # Scale down to 6 cookies and try again.
    ok6, _ = book.can_make("cookies", servings=6)
    print(f"\ncan make 6 cookies? {ok6}")
    if not ok6:
        for line in book.shopping_list("cookies", servings=6):
            print(f"  still need {line.quantity} {line.unit} of {line.ingredient_id}")

    # Top up the pantry, then actually make.
    book.add_stock("sugar", 100, "g")
    book.stock("egg", 6, "whole")
    book.make("cookies", servings=12)
    print("\nafter making 12 cookies, pantry:")
    for iid, item in sorted(book.pantry.items()):
        print(f"  {iid}: {item.quantity} {item.unit}")

File: go/classes/recipe_manager.go
// Recipe Creator and Manager — compositional entities, scaling, pantry checks, shopping lists.
package classes

import (
	"errors"
	"fmt"
)

// Qty is an exact rational quantity. Keep both scaling by 1/3 and integer
// ingredients like "2 eggs" representable without floating point drift.
type Qty struct {
	Num int64
	Den int64
}

// NewQty builds a reduced rational.
func NewQty(num, den int64) Qty {
	if den == 0 {
		panic("denominator cannot be zero")
	}
	if den < 0 {
		num, den = -num, -den
	}
	g := qtyGcd(absInt(num), den)
	if g == 0 {
		g = 1
	}
	return Qty{num / g, den / g}
}

// QtyWhole builds Qty{n, 1}.
func QtyWhole(n int64) Qty { return NewQty(n, 1) }

func qtyGcd(a, b int64) int64 {
	for b != 0 {
		a, b = b, a%b
	}
	if a < 1 {
		return 1
	}
	return a
}

func absInt(x int64) int64 {
	if x < 0 {
		return -x
	}
	return x
}

// Add returns q + r.
func (q Qty) Add(r Qty) Qty { return NewQty(q.Num*r.Den+r.Num*q.Den, q.Den*r.Den) }

// Sub returns q - r.
func (q Qty) Sub(r Qty) Qty { return NewQty(q.Num*r.Den-r.Num*q.Den, q.Den*r.Den) }

// Mul returns q * r.
func (q Qty) Mul(r Qty) Qty { return NewQty(q.Num*r.Num, q.Den*r.Den) }

// Less reports q < r via cross-multiplication.
func (q Qty) Less(r Qty) bool { return q.Num*r.Den < r.Num*q.Den }

// IsZero reports whether q equals zero.
func (q Qty) IsZero() bool { return q.Num == 0 }

// Ingredient is a stockable good.
type Ingredient struct {
	ID          string
	Name        string
	DefaultUnit string
}

// RecipeLine is one ingredient requirement.
type RecipeLine struct {
	IngredientID string
	Quantity     Qty
	Unit         string
}

// Scaled returns a copy with quantity multiplied by factor.
func (l RecipeLine) Scaled(factor Qty) RecipeLine {
	return RecipeLine{IngredientID: l.IngredientID, Quantity: l.Quantity.Mul(factor), Unit: l.Unit}
}

// Recipe is a structured compound entity.
type Recipe struct {
	ID     string
	Name   string
	Lines  []RecipeLine
	Serves int64
	Steps  []string
}

// Scaled returns a pointwise-scaled recipe.
func (r Recipe) Scaled(factor Qty) Recipe {
	lines := make([]RecipeLine, len(r.Lines))
	for i, l := range r.Lines {
		lines[i] = l.Scaled(factor)
	}
	steps := append([]string{}, r.Steps...)
	return Recipe{ID: r.ID, Name: r.Name, Lines: lines, Serves: r.Serves, Steps: steps}
}

// PantryItem is a stocked amount in a specific unit.
type PantryItem struct {
	Quantity Qty
	Unit     string
}

// ShortageReason categorizes why a line can't be satisfied.
type ShortageReason string

const (
	ShortageInsufficient ShortageReason = "insufficient"
	ShortageUnitMismatch ShortageReason = "unit_mismatch"
	ShortageNotStocked   ShortageReason = "not_stocked"
)

// Shortage reports one line's gap between required and available.
type Shortage struct {
	IngredientID string
	Required     Qty
	Available    Qty
	Unit         string
	Reason       ShortageReason
}

// Recipe errors.
var (
	ErrUnknownIngredient = errors.New("unknown ingredient")
	ErrUnknownRecipe     = errors.New("unknown recipe")
	ErrUnitMismatch      = errors.New("unit mismatch")
	ErrInsufficientStock = errors.New("insufficient stock")
)

// RecipeBook holds ingredients, recipes, and the pantry.
type RecipeBook struct {
	Ingredients map[string]Ingredient
	Recipes     map[string]Recipe
	Pantry      map[string]*PantryItem
}

// NewRecipeBook builds an empty book.
func NewRecipeBook() *RecipeBook {
	return &RecipeBook{
		Ingredients: make(map[string]Ingredient),
		Recipes:     make(map[string]Recipe),
		Pantry:      make(map[string]*PantryItem),
	}
}

// AddIngredient registers an ingredient.
func (b *RecipeBook) AddIngredient(i Ingredient) error {
	if _, ok := b.Ingredients[i.ID]; ok {
		return fmt.Errorf("%w: ingredient %s", ErrDuplicate, i.ID)
	}
	b.Ingredients[i.ID] = i
	return nil
}

// AddRecipe registers a recipe; every referenced ingredient must exist.
func (b *RecipeBook) AddRecipe(r Recipe) error {
	if _, ok := b.Recipes[r.ID]; ok {
		return fmt.Errorf("%w: recipe %s", ErrDuplicate, r.ID)
	}
	for _, l := range r.Lines {
		if _, ok := b.Ingredients[l.IngredientID]; !ok {
			return fmt.Errorf("%w: %s", ErrUnknownIngredient, l.IngredientID)
		}
	}
	b.Recipes[r.ID] = r
	return nil
}

// Stock sets (not adds) the pantry amount for an ingredient.
func (b *RecipeBook) Stock(ingredientID string, quantity Qty, unit string) error {
	if _, ok := b.Ingredients[ingredientID]; !ok {
		return fmt.Errorf("%w: %s", ErrUnknownIngredient, ingredientID)
	}
	b.Pantry[ingredientID] = &PantryItem{Quantity: quantity, Unit: unit}
	return nil
}

// AddStock increases pantry stock; fails on unit mismatch.
func (b *RecipeBook) AddStock(ingredientID string, quantity Qty, unit string) error {
	if _, ok := b.Ingredients[ingredientID]; !ok {
		return fmt.Errorf("%w: %s", ErrUnknownIngredient, ingredientID)
	}
	if p, ok := b.Pantry[ingredientID]; ok {
		if p.Unit != unit {
			return fmt.Errorf("%w: %s pantry in %s, adding %s", ErrUnitMismatch, ingredientID, p.Unit, unit)
		}
		p.Quantity = p.Quantity.Add(quantity)
		return nil
	}
	b.Pantry[ingredientID] = &PantryItem{Quantity: quantity, Unit: unit}
	return nil
}

// Scale returns a scaled copy of a recipe.
func (b *RecipeBook) Scale(recipeID string, factor Qty) (Recipe, error) {
	r, ok := b.Recipes[recipeID]
	if !ok {
		return Recipe{}, fmt.Errorf("%w: %s", ErrUnknownRecipe, recipeID)
	}
	return r.Scaled(factor), nil
}

// requiredLines returns the scaled line set for the requested servings.
// Pass servings=0 to use the recipe's native serves value.
func (b *RecipeBook) requiredLines(recipeID string, servings int64) ([]RecipeLine, error) {
	r, ok := b.Recipes[recipeID]
	if !ok {
		return nil, fmt.Errorf("%w: %s", ErrUnknownRecipe, recipeID)
	}
	if servings == 0 || servings == r.Serves {
		return append([]RecipeLine{}, r.Lines...), nil
	}
	factor := NewQty(servings, r.Serves)
	out := make([]RecipeLine, len(r.Lines))
	for i, l := range r.Lines {
		out[i] = l.Scaled(factor)
	}
	return out, nil
}

// CanMake reports (ok, shortages). Pass servings=0 for the native recipe size.
func (b *RecipeBook) CanMake(recipeID string, servings int64) (bool, []Shortage, error) {
	lines, err := b.requiredLines(recipeID, servings)
	if err != nil {
		return false, nil, err
	}
	var shortages []Shortage
	for _, l := range lines {
		p, ok := b.Pantry[l.IngredientID]
		switch {
		case !ok:
			shortages = append(shortages, Shortage{
				IngredientID: l.IngredientID, Required: l.Quantity, Available: QtyWhole(0),
				Unit: l.Unit, Reason: ShortageNotStocked,
			})
		case p.Unit != l.Unit:
			shortages = append(shortages, Shortage{
				IngredientID: l.IngredientID, Required: l.Quantity, Available: QtyWhole(0),
				Unit: l.Unit, Reason: ShortageUnitMismatch,
			})
		case p.Quantity.Less(l.Quantity):
			shortages = append(shortages, Shortage{
				IngredientID: l.IngredientID, Required: l.Quantity, Available: p.Quantity,
				Unit: l.Unit, Reason: ShortageInsufficient,
			})
		}
	}
	return len(shortages) == 0, shortages, nil
}

// ShoppingList returns the gap-report: required minus available, per line.
// Only lines needing more appear.
func (b *RecipeBook) ShoppingList(recipeID string, servings int64) ([]RecipeLine, error) {
	lines, err := b.requiredLines(recipeID, servings)
	if err != nil {
		return nil, err
	}
	var out []RecipeLine
	for _, l := range lines {
		p, ok := b.Pantry[l.IngredientID]
		if !ok || p.Unit != l.Unit {
			out = append(out, l)
			continue
		}
		deficit := l.Quantity.Sub(p.Quantity)
		if !deficit.IsZero() && !deficit.Less(QtyWhole(0)) {
			out = append(out, RecipeLine{IngredientID: l.IngredientID, Quantity: deficit, Unit: l.Unit})
		}
	}
	return out, nil
}

// Make atomically consumes the ingredients required by the scaled recipe.
// All lines must be satisfiable; no partial consumption occurs on failure.
func (b *RecipeBook) Make(recipeID string, servings int64) error {
	ok, shortages, err := b.CanMake(recipeID, servings)
	if err != nil {
		return err
	}
	if !ok {
		return fmt.Errorf("%w: %d shortage(s) on %s", ErrInsufficientStock, len(shortages), recipeID)
	}
	lines, _ := b.requiredLines(recipeID, servings)
	for _, l := range lines {
		b.Pantry[l.IngredientID].Quantity = b.Pantry[l.IngredientID].Quantity.Sub(l.Quantity)
	}
	return nil
}

// RecipesUsing returns every recipe referencing an ingredient.
func (b *RecipeBook) RecipesUsing(ingredientID string) []Recipe {
	var out []Recipe
	for _, r := range b.Recipes {
		for _, l := range r.Lines {
			if l.IngredientID == ingredientID {
				out = append(out, r)
				break
			}
		}
	}
	return out
}

// IngredientsUsed returns the set of ingredients referenced by any recipe.
func (b *RecipeBook) IngredientsUsed() map[string]struct{} {
	out := make(map[string]struct{})
	for _, r := range b.Recipes {
		for _, l := range r.Lines {
			out[l.IngredientID] = struct{}{}
		}
	}
	return out
}
